### PR TITLE
[SPARK-6478] New RDD.pipeWithPartition method

### PR DESCRIPTION
**Pull request by redbaron**
This method allows building command line args and process environement map
using partition as an argument.

Use case for this feature is to provide additional informatin about the partition to spawned application in case where partitioner provides it (like in cassandra connector or when custom partitioner/RDD is used).

Also it provides simpler and more intuitive alternative for printPipeContext function.

### CODE REVIEW

**Comment by JoshRosen** [comment]
I'm a little hesitant to want to add a new `withPartition` or `withSplit`-like method, since we've been deprecating those in favor of using things like TaskContext.  Can you address your use-case with `TaskContext.get` and `printPipeContext`?  For example, how about this:

```scala
myRDD.pipe(
   command="....",
   printPipeContext = (p => p("PARTITION=" + TaskContext.get.partitionId())
```

Is the problem that you want to be able to store the partition as part of the command or environment?  If so, then maybe we could generalize this so that the function is invoked with a TaskContext instead of a Partition (in other words, change it to pipeWithContext).

**Comment by redbaron** [comment]
Thanks for taking a look into this PR.

First about `printPipeContext`, using this function is really awkward, because that is not how data processors are normally written. Usually they suck data from stdin, spit it out to stdout and their behaviour controlled with command line options and/or env variables. `printPipeContext` prepends non-data information to a data in a single stream, this adds unnecessary complication as data processor must be either wrapped into a shell script,which consumes just right ammount of input, figures out runtime switches for data processor and then spawns it, or data processor itself must be (re-)written to be aware of this feature. Compare it to use of `pipeWith*` functions, when it all comes naturally in the form of built up command or env vars map.

Next thing is overall deprecation of `withPartition` commands, I noticed this trend and can't say it makes me happy. It is because assuming that the only usefull bit of Partition is it's index, but that is not always the case. Take for example spark cassandra driver, it has it's own partitioner which inherits Partition and contains rich information on the current subset of data, like range of vnodes being processed. Same case can be with any inhouse partitioners, which not only split the data, but also used to deliver metadata about each split right down to executors.

In short what I am trying to achieve is to get an option to access Partition instance while processing given partition.  TaskContext provides just partitionId and unless there is a way to get partition instance using that id, it doesn't solve my problem. I just briefly checked the code and it seems that in places where `TaskContextImpl` is instantiated no Partition instances are available and it would require some plumbing just to deliver it there.

**Comment by rxin** [comment]
In general we have avoided exposing Partition's directly to users. Since this can be implemented as a helper function outside Spark, maybe this doesn't need to go into the core API itself? I'm just worried about adding an API for a niche use case.

**Comment by JoshRosen** [comment]
> Next thing is overall deprecation of withPartition commands, I noticed this trend and can't say it makes me happy.

Actually, I realize now that I was a little inaccurate earlier: the `mapPartitionsWithContext` and `mapPartitionsWithSplit` interfaces were being deprecated, with the former being replaced by TaskContext.get, and the latter being replaced by `mapPartitionsWithIndex` (but unnecessary if you have TaskContext).  Hence my preference towards allowing TaskContext to be used in more place rather than adding new methods that thread through a single piece of information that it provides.

**Comment by JoshRosen** [comment]
> In general we have avoided exposing Partition's directly to users.

Users can get access to Partition objects by implementing custom RDDs, so I suppose a third-party library could expose something like `mapPartitionsWithPartition`, and then use that to construct arguments to a ProcessBuilder, etc.

> Since this can be implemented as a helper function outside Spark, maybe this doesn't need to go into the core API itself?

In retrospect, most of the PipedRDD enhancements probably should have gone into libraries rather than core, since the current API is already kind of unwieldy.

**Comment by redbaron** [comment]
I see your concerns. Would you be happy if I do all the plumbing to make Partition instance available in TaskContext and then make `pipeWithContext` method available?

**Comment by redbaron** [comment]
Guys, so what do you think? There is no easy way to achieve same goal without copy-pasting a whole PipeRDD into a new one.

For a moment I thought I can get away with:
```
class PPipeRDD[T: ClassTag]( prev: RDD[T],
                             cmdFunc: (Partition) => Seq[String],
                             envFunc: (Partition) => Map[String, String])
  extends RDD[String](prev) {

  override protected def getPartitions =  prev.partitions
  override def compute(split: Partition, context: TaskContext) =
    prev.pipe(cmdFunc(split),envFunc(split)).compute(split, context)
}
```

But then I was caught by SPARK-5063. So there is no easy way to get it without changing spark itself.

**Comment by JoshRosen** [comment]
Hey @redbaron, it looks like this has gone stale. Would you mind closing this PR for now in order to de-clutter the review queue? Thanks!
